### PR TITLE
proxy HTTP basic auth

### PR DESCRIPTION
**Pull request by sunthera**
Added header for basic proxy authorization

### CODE REVIEW

File: socks.py [hunk 1]
```diff
@@ -625,8 +626,17 @@ def _negotiate_HTTP(self, dest_addr, dest_port):
         # If we need to resolve locally, we do this now
         addr = dest_addr if rdns else socket.gethostbyname(dest_addr)
 
-        self.sendall(b"CONNECT " + addr.encode('idna') + b":" + str(dest_port).encode() +
-                     b" HTTP/1.1\r\n" + b"Host: " + dest_addr.encode('idna') + b"\r\n\r\n")
+        http_headers = [
+            b"Connect %s:%s HTTP/1.1" % (addr.encode('idna'), str(dest_port).encode()),
```
Comment (vadmium) [inline socks.py]: Adding b". . ." % . . . formatting would break Python 3.4 compatibility, which I assume still works

File: socks.py [hunk 2]
```diff
@@ -625,8 +626,17 @@ def _negotiate_HTTP(self, dest_addr, dest_port):
...
+            b"Host: %s" % dest_addr.encode('idna')
+        ]
+
+        if username and password:
+            http_headers.append(b"Proxy-Authorization: basic %s" % str(b64encode("%s:%s" % (username, password))))
```
Comment (vadmium) [inline socks.py]: b64encode() does not accept text strings (in Python 3). You will have to explicitly encode the string.

In Python 2, %s is redundant with str(), and in Python 3.5, str() on a byte string is a bad idea, plus I don’t think %s will accept the result.

Comment (sunthera) [inline socks.py]: Yes, you are right, I don't realize about Python 3 compatibility.
Updating:
```%```-formatting with ```b"..." + b"..."```
```b64encode()``` update with ```b64encode(username + b":" + password)```
Will be ok?

Comment (vadmium) [inline socks.py]: Yes I think so. It looks like the username and password have already been encoded to bytes when the proxy settings are set up (I didn’t realize that in my previous comment).

File: socks.py [hunk 3]
```diff
@@ -625,8 +626,17 @@ def _negotiate_HTTP(self, dest_addr, dest_port):
         # If we need to resolve locally, we do this now
         addr = dest_addr if rdns else socket.gethostbyname(dest_addr)
 
-        self.sendall(b"CONNECT " + addr.encode('idna') + b":" + str(dest_port).encode() +
-                     b" HTTP/1.1\r\n" + b"Host: " + dest_addr.encode('idna') + b"\r\n\r\n")
+        http_headers = [
+            b"Connect " + addr.encode('idna') + b":" + str(dest_port).encode() + b" HTTP/1.1",
```
Comment (vadmium) [inline socks.py]: The request method is case-sensitive. You should keep it uppercase.

Comment (sunthera) [inline socks.py]: Yes, inattentively. Typo

**Pull request merged**